### DISCUSSION TITLE
Filament v3: App framework for the TALL stack

### DISCUSSION THREAD

**Discussion by danharrin** [filamentphp/filament#4400] [General]
Ever since Filament launched I've heard of users using our "admin panel" (filament/filament) product for building their entire app. They enjoy the admin panel CRUD features and prebuilt authentication and layout. Their users have access to the dashboard, and this significantly cuts development time. We've also had feature requests to take this further and provide billing integrations for SaaS apps.

Until today, I haven't discouraged this behaviour and said that it was a possible use case for our admin panel. However, I did recommend that users build their apps outside the admin panel and use the form builder, table builder and notifications packages. There are a few reasons for this:

- The admin panel has a page layout and design which is customisable but not completely flexible. I believe that brands should have personality shine through within their app, and the admin panel doesn't always allow this to the fullest extent.
- Out of the box, it's only possible to have one set of resources / pages / dashboard widgets that all users have access to. Granted, you can use authorisation to control access of these to different user groups, but when two different groups have completely different resources then there is no separation between these on the code-level.
- Many SaaS apps have the concepts of "team multitenancy" and billing. This allows multiple groups of users to use the admin panel, switch groups, and collaborate on their own database records. This wasn't something that was easily built into Filament.
- You wouldn't be able to easily have a separate admin section to the main app anymore.

In truth, "admin panel" is purely a marketing term for that package. There are no references of the word "admin" within the codebase of Filament. It's just Filament, it's own thing.

Today, I was discussing this topic further with the rest of the Filament team. It actually started with Dennis:

![Message from Dennis containing: "Dan: Don't use the admin panel for your apps. Me: Could I squeeze this existing thing into the admin panel"](https://user-images.githubusercontent.com/41773797/194403868-eb43b559-dd9f-4d0f-a557-2876f2cfa783.png)

Zep prompted discussion on why I thought this way, and I kinda started to come around to the idea of using the "admin panel" for full apps. It makes sense - so many apps are CRUD-orientated, and many developers want to save time by not needing to build a layout and page navigation and everything.

But even if I thought this way, I still had many concerns (outlined above) as to why I would not use the "admin panel" for such a project. So I thought about it some more and had a discussion. We also made a [Twitter poll](https://twitter.com/filamentphp/status/1577986230112817152), as if *anyone* could convince me this was a good idea, it would be actual users.

The majority of users we asked on Twitter said that they use Filament to build apps, not just admin panels. I didn't really expect that result, and it kinda solidified that I was thinking in the wrong way about this topic.

I jumped on a call with some of the Filament team and we talked about possibilities some more. Here's what we discussed.

## Proposal

Filament will be a full framework built on the TALL stack for building apps. Out of the box, it acts like an admin panel, so there's nothing to worry about if you only need that. But for those who want to, they can transform it into a multi-context, multi-tenant, billable SaaS product, with a separate admin area.

### Ecosystem overview for v3

The following UI components will be installable outside of `filament/filament` and usable in any TALL stack app:

- `filament/forms` - form builder
- `filament/tables` - table builder
- `filament/notifications` - notifications manager
- `filament/widgets` - dashboard widgets like charts and stats cards [NEW]
- `filament/actions` - easy modals and slide-overs with forms [NEW]

`filament/filament` will be advertised as an "app framework" or "app builder" or "admin panel and app framework", and uses all of the individual UI components. 

### Multi-context

Succeeding the popular [multi-context plugin](https://github.com/artificertech/filament-multi-context) (by @coleshirley) for the admin panel, Filament will be able to split itself into different "contexts". Each "context" is definable in a service provider. Each has its own path (`/admin`, `/app`, `/console` etc.), and has it's own CRUD resources, pages, dashboard widgets, etc. You can decide which Laravel authentication guards handle each context, or if the same auth guard should authorize different users to different contexts.

### Teams

You may allow a context to handle "teams". Teams are groups of users who collaborate together on the same database records. It's a form of multi-tenancy, but also involves UI to switch between teams or create a new team. Contexts don't handle teams by default, but you can configure that behaviour in the service provider.

### Billing

Billing is an important part of SaaS. Filament will ship with adapters that can be used to start subscriptions, managing payments, and more. It will also integrate with teams.  We will release a first-party [Laravel Spark](https://spark.laravel.com) adapter.

And of course you could create your own adapters to connect to a custom billing system.

### Theming

We will dramatically improve the ability to customise the appearance of Filament. We will sell themes for Filament. These can change how your Filament contexts look. You can make your own themes if you wish. We might even offer exclusive themes that can only be purchased once, and basically sell you the complete brand materials.

## Notes

Existing admin panel users will not have their experience affected at all.

No, we are not recommending that everyone uses Filament to build every app in existence, just the ones that make sense.

All of my previous concerns are invalid if we implement these features.

**Comment by coleshirley** [comment]
Love it! I'm proud of my multi-context package but it is very 'hacky' to add context functionality as a plugin. Super excited for all the v3 features

**Reply by danharrin** [reply]
Thank you for your service - the community loves it

**Reply by rich06** [reply]
@coleshirley - your package sound awesome :) is it still supported?

**Reply by Wulfheart** [reply]
https://github.com/artificertech/filament-multi-context

**Reply by rich06** [reply]
Thanks - I only asked as there's 4 open issues dating back to Jan with no response.

**Comment by AyoobMH** [comment]
This is just awesome, truly appreciate your time and efforts, would love to contribute to Filament in the near future.

**Comment by mahfoudfx** [comment]
I don't know if this were the initial vision or it's new, but what about to include an API generator?!

**Reply by zepfietje** [reply]
Thanks for the idea! This has been suggested before, but we don't see this fit in the Filament core. Would definitely be a good plugin though!

**Reply by mahfoudfx** [reply]
I was wondering if it could be a separate component, something like filament/apis, if a plugin, it should be officially maintained/supervised to fit the core sources.

**Reply by zepfietje** [reply]
I think we'd rather see this as a community plugin for now.

**Reply by Uolsen** [reply]
I like that idea. It will be complete package if API will be there too 🤔

**Comment by amaurycid** [comment]
I've tried many admin-panel solutions out there... Filament so far is unmatched. The community thanks you for your efforts and concerns bringing this excellent solution.

**Reply by romaldyminaya** [reply]
@danharrin I agree with all of your proposals. Thanks 🙏🏼

**Comment by brianakidd** [comment]
This is great.  The TALL stack, and especially Filament, are the best things going on in the Laravel community IMO.  Glad to see maintainers generate income from their projects  - its great for sustainability and good for all of us in the long run.  Thanks

**Reply by brianakidd** [reply]
I've started on a small SaaS using v2 because I can't wait for v3.  With v3 including multi-context, teams, etc., will I be able to upgrade to v3 or will it be easier to start with a new v3 project?  The Laravel project was scaffolded using jetstream for livewire with the --jet and --teams flags.  Thanks

**Reply by danharrin** [reply]
You won't need Jetstream anymore but I doubt you will need to start a new project

**Reply by brianakidd** [reply]
> You won't need Jetstream anymore but I doubt you will need to start a new project

Do you foresee having a "team" model where a user can belong to many teams as in jetstream or some other way for managing tenants?

EDIT: Sorry I went back and reread your post and it does seem to indicate a user can belong to many teams.  Thanks

**Reply by danharrin** [reply]
Yeah, and thats already built and ready for release

**Reply by bfiessinger** [reply]
will it be possible to use an alternate team model?

**Reply by danharrin** [reply]
Absolutely. You can use any tenant model you wish. Even the User.

**Reply by bfiessinger** [reply]
@danharrin can't wait to upgrade to filament v3. Love all the work done so far! ❤️

**Comment by MACscr** [comment]
I dont see a downside to any of the proposal. Sounds great.

**Comment by ralphjsmit** [comment]
Awesome, love the new direction! I think that the "themes" stuff is really good. Thanks for writing this up 🙌❤️

**Comment by lukas-frey** [comment]
This is briliant. Writing filament apps is such an enjoyable experience, can't wait for this!

**Comment by martin-ro** [comment]
Me every time working with Filament: "🤔 Wonder if I can use this in the front-end"

**Reply by zepfietje** [reply]
We've heard you thinking. 😏

**Comment by blite** [comment]
Love this, very excited for the teleport features.

**Comment by invaders-xx** [comment]
I already use it for frontend/backend and other contexts using the great @coleshirley package! And I enjoy the direction Filament takes!
Long live Filament!

**Reply by zepfietje** [reply]
Thanks! That proves it's the right direction. 🚀

**Comment by vv-ac** [comment]
This is truly what I was hoping for but did not allow myself to ask for. 🐱‍🏍

**Comment by nbejansen** [comment]
Great direction! I just finished a package which gives Laravel Jetstream a Filament inspired UI (including team switching and Stripe Customer Portal integration). Maybe I can contribute!

Preview of the package I developed:
https://www.loom.com/share/52d22070c161458aa7a4ec948a3910b2

**Reply by AyoobMH** [reply]
Looks neat 👌

**Reply by notdone-au** [reply]
Looks good planning to make this specific package public ?

**Comment by Pen-y-Fan** [comment]
FATALL stack is getting better and better 🎉
(Filament Admin Tailwind Alpine Laravel Livewire)

**Comment by mikrosmile** [comment]
I would like to pint some thoughts why I choose filament for core app

First I was planning to build my apps using C# and faced that every step I had to create an action / method to do something (simple create / edit records)

Then I was looking for something like CRUD panel, actually I found some on C# (Razr) but also faced limitations with customization and the way how I want some action to be performed

And finally I decided to return to language I know, which is PHP, where I found a great framework Laravel, it fits my needs on 100% and I was searching for an admin panel for that ... and found filament 

Since the first day I never consider filament as "admin" panel or "backend", because it has a loft of functionality, the design also fits my UX language & preferences, I like simplicity, yet powerful options to implement all I want, and since that I said myself, this will be my core app, of course I'll build some "external" features from outside filament, but the main core functionality will be filament still. 

I like the way where it goes so far, and SaaS billing is what I'm actually waiting for :)

**Reply by martin-ro** [reply]
Hell of a pivot from C# to PHP, loving it 😅

**Reply by Wulfheart** [reply]
Language-wise I prefer C#, but Filament is so damn good it is a reason to use Laravel on its own. 😍

**Comment by invaders-xx** [comment]
Any ideas on the v3 roadmap? availability ?

**Reply by danharrin** [reply]
No idea, I don't have a date from Caleb as to when Livewire v3 will be at a beta stage where we can work on upgrading

**Comment by mikrosmile** [comment]
> The following UI components will be installable outside of `filament/filament` and usable in any TALL stack app:

Also, would be good to have filament/pages (i.e "Custom pages") as a package to use from outside :)

**Reply by danharrin** [reply]
but custom pages don't really do anything on their own?

**Reply by mikrosmile** [reply]
True, but it's already has its own layout, you can place widgets, header, footer, its also way simple to use multiple tables in one place (which was asked by many), so when i wanted to have several tables in one page I created a custom one and placed widgets there.. 

```
protected function getHeaderWidgets(): array
    {
        return [
            Countries::class,
            Regions::class,
            Cities::class
        ];
    }

```

<img width="1307" alt="Снимок экрана 2022-10-08 в 19 24 50" src="https://user-images.githubusercontent.com/2136612/194717513-9d181672-9435-469a-89f4-73ccc158b7d6.png">

And it is just a small portion of what `filament/Pages/Page` can do. And ways of customization of that page is beyond my imagine :) 
So I thought it would be also great to have something from outside (and because its just already exists)

**Reply by mikrosmile** [reply]
So I imagine filament/pages is sort of dashboard from outside the filament with all the goodies of filament components :)

**Reply by mikrosmile** [reply]
Or `filament/dashboard ` if you like it more

**Reply by zepfietje** [reply]
Isn't that what `filament/widgets` will be? 🙂

**Reply by mikrosmile** [reply]
Well, maybe it's the same 😃

**Comment by ArtMin96** [comment]
:fire: Is it possible to add an option for the table to work with the data received from the API?

**Reply by romaldyminaya** [reply]
Many people are requesting this feature, it would be very helpful

**Reply by drhoussem** [reply]
I don't this would a good idea! In filament you get filtering, pagination, search, actions, ordering..out of the box while you would need to have different endpoints for each of this features.

**Reply by ArtMin96** [reply]
> I don't this would a good idea! In filament you get filtering, pagination, search, actions, ordering..out of the box while you would need to have different endpoints for each of this features.

If you have a use case where you need to get data from the API and display it, how would you do it?

**Reply by mikrosmile** [reply]
```
   
private function showOptions($inn,$kpp,bool $branch)
    {
            if ($branch){
                $dadata = (new DaDataCompany)->id($inn, 20,$kpp ?? null,BranchType::BRANCH);
            }else{
                $dadata = (new DaDataCompany)->id($inn, 20,$kpp ?? null);
            }
            $array = $dadata['suggestions'];
            $this->dataMassive = $array ?? null;
            $data = collect([]);
            for ($i = 0; $i < count($array); $i++) {
                $name = $array[$i]['value'];
                $data2 = $array[$i]['data'];
                $inn = $data2['inn'];
                $kpp = $data2['kpp'];
                $innKPP = ' (ИНН '.$inn.' / КПП '.$kpp.')';
                $value = $name.''.$innKPP;
                $data->push($value);
            }
            return $data ?? null;
    }

```

this simple function loads data from api, displays options in select and populate form data with ->fill method ... simple as ..
Also, I fill some data in form, click button which gets data from form to specify my request to API

**Reply by ArtMin96** [reply]
> ```
>   
> private function showOptions($inn,$kpp,bool $branch)
>    {
>            if ($branch){
>                $dadata = (new DaDataCompany)->id($inn, 20,$kpp ?? null,BranchType::BRANCH);
>            }else{
>                $dadata = (new DaDataCompany)->id($inn, 20,$kpp ?? null);
>            }
>            $array = $dadata['suggestions'];
>            $this->dataMassive = $array ?? null;
>            $data = collect([]);
>            for ($i = 0; $i < count($array); $i++) {
>                $name = $array[$i]['value'];
>                $data2 = $array[$i]['data'];
>                $inn = $data2['inn'];
>                $kpp = $data2['kpp'];
>                $innKPP = ' (ИНН '.$inn.' / КПП '.$kpp.')';
>                $value = $name.''.$innKPP;
>                $data->push($value);
>            }
>            return $data ?? null;
>    }
> ```
> 
> this simple function loads data from api, displays options in select and populate form data with ->fill method ... simple as .. Also, I fill some data in form, click button which gets data from form to specify my request to API

We are talking about working with table.

**Comment by andrewdwallo** [comment]
Glad to here this! I was actually having a hard time trying to implement API's to be used in the admin panel for my ERP Management System I was trying to build and about a week ago I stopped using Filament for building what I needed because it just wouldn't work. But after seeing this I am truly excited because Filament may be one of the only "Admin" panel crud generators that was very extendable up until the point of my certain use case of a SaaS App. One of the only ones that has great UI and Design and is actually modern looking. Every other package out there that may be considered similar to that of Filament has a terrible design and is simply outdated... 
THIS WILL BE A GAME CHANGER!!! I hope to collaborate!

**Reply by margarizaldi** [reply]
And the most disturbed party in this competition is: yes, the "official" one...

**Comment by eugenefvdm** [comment]
I'm loving this direction and I think the billing integration is a game changer.. The point to me is that Laravel is a great SaaS platform, but everything is a bit all over the show. You get to use the framework, then you have to get the back office, then you have to get the billing system (and maybe pay for it), then you have to build out a front-end. These things are not so closely integrated at the moment. But if you're working towards a SaaS you end up spending huge resources building out all those areas. Something more closely coupled makes a lot of sense, less technology to learn and a quicker time to market, which is crucial for success.

**Comment by mutuku** [comment]
Please add support for Paddle (https://laravel.com/docs/9.x/cashier-paddle) under billing. It is popular in countries where Stripe is not supported.

**Reply by danharrin** [reply]
We will integrate with Laravel Spark  which supports Paddle

**Reply by margarizaldi** [reply]
> We will integrate with Laravel Spark  which supports Paddle

@danharrin I hope this should be a plugin or module, since payment is too opinionated... And i think considering it as paid option is not too bad... They're doing business, why not? 😉

**Reply by danharrin** [reply]
Today I completed the Spark integration. It is a separate plugin, and you can build your own integration if you wish.

**Comment by blite** [comment]
One thing I think should really be standard is https://github.com/HollyIT/filament-static-assets (note I don't even use this package myself). But the idea behind it that the static files should serve from public is imo something that needs to be on the roadmap.  I would 100% understand reasons for delaying it (letting vite settle etc), but long term the static assets should be serving from nginx, and/or should have a path for pushing to cdns directly.  There may be alternatives that mitigate the importance of this such as using cloudflare, but I suspect you would find setting up cloudflare rules would be easier if the static files where in standard folders.

**Reply by zepfietje** [reply]
This will be fixed by the new asset management system in v3 which copies package assets into your  application's `public` folder.

**Comment by thyseus** [comment]
One could take a look at Gii, the a "Code Generator" that ships with the Yii Framework:

https://www.yiiframework.com/doc/guide/2.0/en/start-gii

This is sort of "filament light". It started as a "code generator" for the whole application, including the possibility to create admin interfaces.
Maybe there are some ideas in there that can be borrowed for filament-v3.

**Reply by ArtMin96** [reply]
In Yii, Gii is the same as a resource generator in filament, just with UI.

[Creating a resource](https://filamentphp.com/docs/2.x/admin/resources/getting-started#creating-a-resource)

**Reply by thyseus** [reply]
Thats true. Only addition to filament that gii provides so far is, that it "guesses" how the form/table looks from the existing sql table when generating the resources. For every column of the sql table a PHP Line is generated, depending on the datatype of the column.

But Gii has more to offer, a deeper look just for collecting ideas shouldn´t hurt.

**Reply by oplanre** [reply]
Filament also does that with the --generate flag on make:filament-resource, though you need doctrine/dbal. check out the docs

**Comment by Wulfheart** [comment]
Every time I read this I hope that Caleb Porzio hurries up so the development for Filament v3 continues because I can't wait to get my hands on it.

**Reply by zepfietje** [reply]
We're actually working hard on Filament v3 already! Ultimately depends on Livewire v3 release date, but we're definitely not fully blocked. 😄

**Reply by Wulfheart** [reply]
Let's summarize it: I am really looking forward to it!

**Comment by imClement** [comment]
Great plans and great work. 

While I understand the v3 makes jetstream unnecessary, 
will they still play nicely within the same app? 

Will the jetstream teams be compatible with the multi-context approach or, whenever possible, for new projects the jetstream teams should be avoided and use the Filament native approach for teams?

**Reply by danharrin** [reply]
Yeah there's no reason to use Jetstream teams if you use our teams. But if you wanted to add a **separate** admin panel onto a Jetstream teams app then you could.

**Reply by imClement** [reply]
Thank you.

I am really tempted to hold the new project start until v3. 

The problem is I don't have any clue  on how much can take (weeks, months).
I understand it is tied to the Livewire release but still no sense of timing. 

Do you have any recommendation of a source to keep an eye on it for timing hints on the Filament or Livewire release?

**Reply by danharrin** [reply]
Caleb's Twitter will be the first place where information is released

**Comment by mrezauli** [comment]
Why not an official plugin/package for cms like this https://filamentphp.com/plugins/sky???

**Reply by notdone-au** [reply]
I just feel there are many use cases for filament why is a CMS important enough to be a core package, compared to all the other things possible with filament?

**Reply by danharrin** [reply]
way too opinionated, CMSes are easy to build with Filament exactly to your requirements. no need for a plugin.

**Reply by romaldyminaya** [reply]
> way too opinionated, CMSes are easy to build with Filament exactly to your requirements. no need for a plugin.

Agree with @danharrin

**Comment by margarizaldi** [comment]
Is there any plan to enable kind of "client-side table"? Which all rows are rendered once, then functionalities like sorting, filtering, etc. will be handled by javascript (similar to DataTable)... I think this will be useful for small tables...

**Reply by zepfietje** [reply]
This would essentially require a total rebuild of the tables package. There's no plans for that.

**Reply by margarizaldi** [reply]
Hmmm alright...

I was wondering that it wouldn't be that hard... Assuming we can return all rows and "include" some nice js table library available out there...

My case: it's very hard to optimize many conditional sums and counts from nested relationships, while every table action (even simply sorting column) will literally re-execute the query again and again...

**Reply by oplanre** [reply]
Can you recommend a js library? All the cool ones i see are mad expensive

**Reply by thyseus** [reply]
https://www.ag-grid.com/ 

the free package has 99% of features a mortal private person ever needs :)

**Comment by ekandreas** [comment]
Yes!

**Reply by nnq1** [reply]
What is the plan to release Filament v3? even an estimation!

**Reply by zepfietje** [reply]
There's no ETA yet. Livewire v3 is probably still a couple of months away, and we depend on that.

**Reply by nnq1** [reply]
thanks

**Reply by imrancoder** [reply]
@zepfietje 
what will be recommendations to follow, if we start new project(multi context) on filament in January 2023
so we can easily switch to v3 after filament V3 release

**Reply by zepfietje** [reply]
There's no specific recommendations. We'll make upgrading as easy as possible.

**Reply by margarizaldi** [reply]
> There's no ETA yet. Livewire v3 is probably still a couple of months away, and we depend on that.

I'm wondering that 2023 will be the year of Splade than Livewire... No offense, but I think splade is interesting to explore, maybe there will be an spa version "filament", who knows... Cheers...

**Reply by danharrin** [reply]
I don't see any benefits of Splade over Livewire v3 really... and Filament v3 will likely use Livewire's new SPA mode anyway

**Reply by danharrin** [reply]
Also Splade still uses traditional Laravel controllers, which wouldnt work for something like Filament as it would close down customisation opportunities. Livewire components are our silver bullet

**Reply by margarizaldi** [reply]
> I don't see any benefits of Splade over Livewire v3 really... and Filament v3 will likely use Livewire's new SPA mode anyway

Ah really? Interesting...

My opinion was influenced by my experience using filament v2, although everything worked really well, fast enough, and beautiful, it still however didn't feel as fast as SPA... Especially I would never use is the native reactive field which is (again, my subjective) not comfortable for users... But **no complain at all**, I've got much much benefits using filament... Love it... 

And I'm sorry if my opinion about splade over livewire was far from the fact, I think I need to change the sentence "I'm wondering that ..." into "I'm **afraid** that..." (not native English speaker anyway)...

**Reply by andrewdwallo** [reply]
You can easily change that by adding a small plugin, and it will be faster than splade... I used splade before but I don't think there is anything that makes it stand out more. It just depends on whether you like using Vue and Inertia more than Livewire and if you are more comfortable with it. Other than that I think Filament's community and the Creators and maintainers of it make it far better in my opinion, and that makes it far more valuable in my book...

**Reply by ekandreas** [reply]
I hope the day never will come when Filament goes backward in tech stack to Inertia and Vue.
That is why we are leaving Nova. 

The idéa of separating the building blocks from the admin is just great!

**Reply by vitorpena** [reply]
> Livewire v3

Hi @andrewdwallo, can you please expand on the plugin you mentioned? Thanks!

**Reply by andrewdwallo** [reply]
@vitorpena https://filamentphp.com/tricks/hotwire-turbo-stimulus-with-vite

**Reply by Sophist-UK** [reply]
I am at the beginning stages of creating an open source Laravel code generator and having to decide whether to use Livewire or Splade (and which Admin panel to use).

From what I know so far, the main benefit of Splade vs. Livewire is that Splade gives you access to a **HUGE** base of open-source Vue components, whereas Livewire components are fewer and many are commercial rather than open source.

So if you are a Laravel back end developer who would rather use existing Vue components than have to create a Livewire equivalent,. I think Splade might have a significant advantage.

Have I understood this correctly or got it completely wrong?

**Reply by Sophist-UK** [reply]
@MACscr Mark

I see you downvoted my comment - is that because I have got this completely wrong or because you just don't like the comment?

**Reply by tpetry** [reply]
You're downvoted because this is a discussion about a package built on Livewire. We're not here to discuss technological differences for your package. You can ask on Reddit, Twitter, Laracasts Forum and many other places.

And you're also comparing apples to oranges. SPLADE is is used by a tiny fraction of developers while Livewire and Inertia.js are the common top dogs together with blade. So event Inertia.js would be better than SPLADE.

But as said: It's the wrong point to start such a discussion!

**Reply by Sophist-UK** [reply]
Ok. Thx.

**Comment by booni3** [comment]
I am excited for this! Especially for greater theming opportunities. I V2, I do find theming a little cumbersome vs. building my own app (using something like Tailwind UI components). Interested to see what ideas you come up with.

**Comment by MACscr** [comment]
Lets not get off topic here, this is a TALL stack project. It wont change. Someone can build a InertiaJS one if they want.

- Mark

**Comment by R00118189** [comment]
Will V3 have decoupled views? I understand https://filamentphp.com/blog/publishing-views-in-laravel shows the dangers of publishing views, but decoupling could give devs more flexibility in regards to how the final product looks

**Reply by Wulfheart** [reply]
Interesting thought. How would zog decouple this?

**Reply by danharrin** [reply]
We are satisfied with customising everything with CSS. Decoupling such a huge part of Filament would make the framework quite unmaintainable for us. I am curious as to what you are not able to do with just CSS as we recommend.

**Reply by R00118189** [reply]
@danharrin, it's less about CSS and more about component design and internal logic. Currently, I create my class extending, say, wizard class and use my customised view file with its internal alpine code. I wonder if it would be possible to add an extra method to overwrite the hardcoded view of the original component. I fully understand the risk you mentioned in the blog post of publishing views. 
Therefore if the new view and its internal logic follow the one you maintain, it is easily switchable from the form definition. So we only add the views we need. This would vastly expand our options for the same functionality with various looks.

**Reply by danharrin** [reply]
Yeah we have that already everywhere in Filament, pass a new ->view() with the name of your custom view

**Comment by tpetry** [comment]
I don't know whether I am a special snowflake with my requirements. I would like to see the `view` pages enhanced. Currently this is only the form in a disabled state which is kind of ok for an admin framework. But when using it as an application I would like (on default) to show the data more appealing than using a form. Like Laravel Nova is doing it that you have a nice details page that is not looking like a form at all.

I know that I could write a special Laravel view for every component but that is really tedious and looks strange as I can't fully replicate the same styling. Having this by default as a new rendering way for every filament component would be nice.

**Reply by danharrin** [reply]
Hey, not a snowflake request at all. We do get it quite alot.

It's highly debated what approach we would take though:
- New method on the Resource class that allows you to define an entirely different structure of data, just for viewing records with. For instance, `details()` with an array of Details components like `TextDetail`, `ImageDetail`, `RepeaterDetail` etc. I like this idea, but I feel like people would be put off if they had to define `form()`, `table()` AND `details()`.
- New set of views for the existing form components. When you're on the view page, these new views are being used, but the structure of the form remains the same. This is the easiest way to implement such a feature, but then you have to stick to the `form()` structure. And it's weird still using a `TextInput` when there's no input.

What do you think?

**Reply by tpetry** [reply]
I prefer the first way. Couldn't there be a simple solution that when you don't want to type everything twice you can call a helper that transforms a form schema to a detail schema?

And when you want to use something more special you define it twice. Its a common problem that I have with Nova. I want data to look differently on the listing and the details which is really cumbersone there (you need to define it twice and hide/show) and not easy to understand if you want to change the ordering for both.

I don't have a big problem with defining it twice if I have (like with forms) more options in the details part to display it who I would like to. Filament is great because it has so many options!

**Reply by Wulfheart** [reply]
I am with @tpetry

**Reply by danharrin** [reply]
I don't think it'd be possible to automatically convert the form schema into a details schema, would end up pretty complex.

I will work on a POC for the new details pages soon.

**Reply by zepfietje** [reply]
We could actually create a new Nova-like method which maps to a basic tables/forms/details configuration? Then if you want full control, you just use the separate methods.

**Reply by lukas-frey** [reply]
Wouldn't it be possible to create new `<Something>Detail` as Components, so that you can use them inside `form()`? Just in case you wanted to mix input fields with detail fields.

**Reply by nikcani** [reply]
> We could actually create a new Nova-like method which maps to a basic tables/forms/details configuration? Then if you want full control, you just use the separate methods.

Would love to see this. I like the way Nova does it, but it is missing a lot of customization if you need it.

**Comment by romaldyminaya** [comment]
@danharrin ,

Have we already mention about the possibility to use the table control from a collection or API instead of an eloquent query? In V2  we've managed a way to do it with a fake model but we need a more standard approach for this.

Thanks in advance.

**Reply by danharrin** [reply]
No plans for that, we'd have to rebuild all our sorting, searching, filtering... not gonna happen when I have no use for such a feature. Sushi is the official way to handle this.

**Reply by romaldyminaya** [reply]
Thanks for  your fast response.

**Reply by margarizaldi** [reply]
A new table "mode" may be feasible... Which render kind of datatables (or some better library) that shares exact same UI with the existing table, maybe just putting "if-else" everywhere to switch livewire behavior...... All functionalities (filter sorts etc.) usually are provided by the library...

**Comment by Jamesking56** [comment]
Is there a planned release date for Filament v3? Waiting for it to build my SaaS

**Reply by danharrin** [reply]
We need a release date for Livewire v3 first

**Reply by Jamesking56** [reply]
@danharrin Ah! Didn't realise this was waiting for Livewire v3

**Comment by Pen-y-Fan** [comment]
[calebporzio](https://github.com/calebporzio) has posted a discussion for  [livewire/livewire When V3? #5585](https://github.com/livewire/livewire/discussions/5585)

> "When V3?"
> Right now I can't give you a date, but I CAN be transparent about the process and my progress, so here's what I would type to everyone who asks if I had time:
> ...

As Filament v3 is waiting for a release date for Livewire v3 it may be worth watching this discussion for an update.

**Comment by MACscr** [comment]
Because the discussion doesn’t belong here.- MarkOn Feb 26, 2023, at 4:26 PM, Sophist ***@***.***> wrote:﻿
@MACscr Mark
I see you downvoted my comment - is that because I have got this completely wrong or because you just don't like the comment?

**Comment by AidasK** [comment]
You can read all about LiveWire v3 release here https://github.com/livewire/livewire/discussions/5585.
It's still not in beta, so I guess we could see livewire to be released in summer and FilamentPhp might be migrated to the livewire v3 a couple months later.

**Comment by elmergustavo** [comment]
in love the proposals, thanks for your efforts👌

**Comment by joshtorres** [comment]
This sounds great! I can't wait to see how we can integrate this with the Catalyst framework as we've discuss if it would be possible to use filament for the entire app several times. I'm looking forward to seeing this come out! 😁

**Comment by Jamesking56** [comment]
Is any of this tagged anywhere in Alpha or Beta form so we can start using it to upgrade to stable later?

**Reply by josefbehr** [reply]
No. But you can already use the 3.x branch of the repo, if you want to. But beware that there is no guarantee of stability or changes not breaking stuff at the moment.

**Comment by cnkbekir** [comment]
> Livewire V3 will be publicly available after my Laracon talk on July 20th.
> -Caleb

https://v3when.com/

**Comment by danharrin** [comment]
As mentioned above, Filament v3 will be released on August 1st! Thanks for being patient everyone.

**Comment by demobiel** [comment]
cool, nice that we have a release day. are there already some docs? i am even stuck at just creating a new project from scratch and trying to add a user. It fails at finding the currentPanel, which I guess is a new feature :) don't want to bother with asking questions, but if there is a doc already available, would be nice to have access to it.

**Reply by zepfietje** [reply]
Using v3 is unsupported as it's still in development.

**Reply by demobiel** [reply]
thanks for the reply! I just wanted to play around with it :) I guess I am just too exited.
